sunscreen/src/compiler.rs

```diff
@@ -398,37 +458,131 @@ impl<T, B> GenericCompiler<T, B> {
     }
 }
 
-impl<T, B> GenericCompiler<T, BoxZkpFn<B>>
-where
-    B: FieldSpec,
-{
-    fn compile_zkp(&self) -> Result<HashMap<String, CompiledZkpProgram>> {
+// This generic impl can contain public methods where the builder remains the same type, or
+// internal methods that are restricted to ZKP-capable builder types.
+impl<T: marker::Zkp, B: FieldSpec> GenericCompiler<T, B> {
+    fn compile_zkp(&self, params: Option<&Params>) -> Result<HashMap<String, CompiledZkpProgram>> {
         let zkp_data = self.data.zkp_data();
 
+        let shared_zkp_programs = zkp_data.shared_zkp_program_fns.iter().map(|prog| {
+            // As long as we've properly maintained invariants internal to the compiler, shared
+            // programs should only be present when params are available.
+            let params = params.expect("no params; please file a bug!").clone();
+            // Note: this is currently unsupported because determining the exact arbitrary
+            // plaintext modulus from the dynamic length of a shared ZKP input in general is not
+            // possible (as x.ilog2() is not injective). We may support this in the future.
+            if !params.plain_modulus.is_power_of_two() {
+                return Err(Error::unsupported(
+                    "Plaintext modulus must be a power of two for ZKP programs with #[shared] arguments.",
+                ));
+            }
+            let result = prog.build(params.plain_modulus)?;
+            let result = zkp::compile(&result);
+            let metadata = ZkpProgramMetadata {
+                params: Some(params),
+                signature: prog.signature(),
+            };
+            let compiled_program = CompiledZkpProgram {
+                zkp_program_fn: result,
+                metadata,
+            };
+            Ok((prog.name().to_owned(), compiled_program))
+        });
         let zkp_programs = zkp_data
             .zkp_program_fns
             .iter()
             .map(|prog| {
-                let result = prog.build()?;
+                let result = prog.build(())?;
                 let result = zkp::compile(&result);
+                let metadata = ZkpProgramMetadata {
+                    params: None,
+                    signature: prog.signature(),
+                };
+                let compiled_program = CompiledZkpProgram {
+                    zkp_program_fn: result,
+                    metadata,
+                };
 
-                Ok((prog.name().to_owned(), result))
+                Ok((prog.name().to_owned(), compiled_program))
             })
+            .chain(shared_zkp_programs)
             .collect::<Result<HashMap<_, _>>>()?;
 
         Ok(zkp_programs)
     }
+}
 
-    fn compile_internal(self) -> Result<Application<T>> {
-        Application::new(HashMap::new(), self.compile_zkp()?)
+impl<B> ZkpCompiler<B>
```

Comment:
It might help to think about the return value of `Compiler::compile`. The idea is to return `Application<Fhe/Zkp>` based on what was passed to the `Compiler`. You start with a blank `Compiler::new()` that can't even `compile()` into anything. You can _always_ add an `fhe_program` to your compiler, and when you do we make sure you get back an `Application<M>` where `M: marker::Fhe`.

It confused me too at first, but then I spent a long time in this file :slightly_smiling_face: